filament/src/main/java/net/fabricmc/filament/task/RemapUnpickDefinitionsTask.java

```diff
@@ -92,58 +126,189 @@ public void execute() {
+							}
+
+							return Type.getDescriptor(field.getType());
+						}
+					});
 					FileUtil.write(output, UnpickUtil.getLfOutput(writer));
 				}
 			} catch (IOException e) {
 				throw new UncheckedIOException(e);
 			}
 		}
+
+		private JarIndex indexJars(Collection<File> jarFiles, MemoryMappingTree mappingTree, int fromM, int classpathM) throws IOException {
+			final List<JarIndex> indexes = new CopyOnWriteArrayList<>();
+			final ThreadLocal<JarIndex> localJarIndex = ThreadLocal.withInitial(() -> {
+				JarIndex index = new JarIndex(new HashMap<>(), new HashMap<>());
+				indexes.add(index);
+				return index;
+			});
+
+			List<ZipFile> zips = new ArrayList<>(jarFiles.size());
+			List<CompletableFuture<Void>> futures = new ArrayList<>();
+
+			try (ExecutorService executor = Executors.newFixedThreadPool(Runtime.getRuntime().availableProcessors())) {
+				for (File jarFile : jarFiles) {
+					zips.add(new ZipFile(jarFile));
+				}
+
+				for (ZipFile zip : zips) {
+					Enumeration<? extends ZipEntry> entries = zip.entries();
+
+					while (entries.hasMoreElements()) {
+						ZipEntry entry = entries.nextElement();
+
+						if (entry.getName().endsWith(".class")) {
+							futures.add(CompletableFuture.runAsync(() -> {
+								ClassReader reader;
+
+								try {
+									reader = new ClassReader(zip.getInputStream(entry));
+								} catch (IOException e) {
+									throw new UncheckedIOException(e);
+								}
+
+								reader.accept(new ClassVisitor(Opcodes.ASM9) {
+									private String className;
+									private String mappedClassName;
+
+									@Override
+									public void visit(int version, int access, String name, String signature, String superName, String[] interfaces) {
+										this.className = name;
+										mappedClassName = mappingTree.mapClassName(name, classpathM, fromM);
+										int slashIdx = name.lastIndexOf('/');
+										String packageName = slashIdx == -1 ? "" : mappedClassName.substring(0, slashIdx);
+
+										localJarIndex.get().classesInPackages
+														.computeIfAbsent(packageName.replace('/', '.'), k -> new ArrayList<>())
+														.add(mappedClassName.replace('/', '.'));
+									}
+
+									@Override
+									public FieldVisitor visitField(int access, String name, String descriptor, String signature, Object value) {
+										String mappedName = name;
+										MappingTree.FieldMapping fieldMapping = mappingTree.getField(className, name, descriptor, classpathM);
+
+										if (fieldMapping != null) {
+											mappedName = fieldMapping.getName(fromM);
+										}
+
+										String mappedDesc = mappingTree.mapDesc(descriptor, classpathM, fromM);
+										localJarIndex.get().fieldDescs.put(mappedClassName.replace('/', '.') + "." + mappedName, mappedDesc);
+										return null;
+									}
+								}, ClassReader.SKIP_CODE);
+							}, executor));
+						}
+					}
+				}
+
+				for (CompletableFuture<Void> future : futures) {
+					future.join();
+				}
```

Comment:
Is this needed? I think closing the exeuctor does this already.

Reply:
Ah, I was under the impression that it could cancel them but you're right